src/main/java/org/truffleruby/stdlib/readline/ConsoleHolder.java

```diff
@@ -73,6 +85,30 @@ public ConsoleHolder() {
         readline.setHistory(history);
     }
 
+    @TruffleBoundary
+    public ConsoleHolder(ConsoleHolder copy, int inFd, int outFd) {
+        in = copy.in.getFd() == inFd ? copy.in : new FileStreamer(inFd);
+        out = copy.out.getFd() == outFd ? copy.out : new FileStreamer(outFd);
+
+        try {
+            readline = new ConsoleReader(in.getIn(), out.getOut());
+        } catch (IOException e) {
+            throw new UnsupportedOperationException("Couldn't initialize readline", e);
+        }
+
+        final ConsoleReader copyReadline = copy.getReadline();
+
+        readline.setHistoryEnabled(copyReadline.isHistoryEnabled());
+        readline.setPaginationEnabled(copyReadline.isPaginationEnabled());
+        readline.setBellEnabled(copyReadline.getBellEnabled());
+
+        currentCompleter = copy.getCurrentCompleter();
+        readline.addCompleter(currentCompleter);
+
+        history = copy.getHistory();
+        readline.setHistory(history);
```

Comment:
It seems simpler to allow updating the `fd` (or File/IO object) of the FileStreamer, than recreating a new ConsoleHolder.

Reply:
I was aiming for immutability in case one thread had a handle to the console holder while another was updating it. The `ConsoleReader` reference is going to have to be recreated because there's no way to update its in & out references.

Or are you advocating for having a delegate `InputStream` that retargets its internal stream reference?

Reply:
Yes, I was advocating to make the InputStream/OutputStream mutable.
It could be achieved by having FileStreamer implement InputStream, and delegate to the FileInputStream as needed.

The SnakeYAML support had adapters from IO to InputStream/OutputStream:
https://github.com/oracle/truffleruby/blob/3b217a9864e07e0108794b7d1e33ced7d375264d/src/main/java/org/truffleruby/core/adapters/InputStreamAdapter.java
https://github.com/oracle/truffleruby/blob/3b217a9864e07e0108794b7d1e33ced7d375264d/src/main/java/org/truffleruby/core/adapters/OutputStreamAdapter.java
https://github.com/oracle/truffleruby/blob/3b217a9864e07e0108794b7d1e33ced7d375264d/src/main/java/org/truffleruby/stdlib/psych/PsychEmitterNodes.java#L142-L149

Reply:
You're right avoiding to change the stream might be better for multi-threaded semantics though, so the current code seems fine too.